README.md
 ![eagle-R]


# What is InspectR?

InspectR gives you insight into quality of your code at a glance. It supports variety of code quality tools. 
When you add `inspectr.json` file to your project directory and use command `inspectr` reports from your test tools will be created and save in RethinkDB. 

The reports will be able in [inspectr-dashboard][dashboard] and it looks like this

![example]

Those are the ones supported now, more under way:

* flake8 python linter
* python unittests
* py.test tests
* coverage.py reports
* eslint javascript linter
* jasmine tests
* mocha tests
* karma-coverage reports


# Quick start (for developers)

1. Install Inspectr globally from pip `pip3 install inspectr`.
2. Run inspectr-dashboard with RethinkDB (go [here][dashboard]).
3. Add inspectr.json to your project root directory. Example:
```
    {
      "project_name": "My Project",
      "rethinkdb_host": "localhost",
	  "rethinkdb_port": 28015,
      
    
      "reporters": [
        {
          "type": "jasmine",
          "command": "karma start karma.conf.js --single-run"
        },
        {
          "type": "mocha",
          "command": "karma start karma.conf.js --single-run"
        }
      ]
    }
```
4. Use command `inspectr` in this directory.
5. If everything went ok you should see results in browser.


# Using inspectr with docker

1. Run RethinkDB, we recommend to use docker-compose in [inspectr-dashboard][dashboard] which is creating RethinkDB container.
2. Add `inspectr.json` to your project root directory (example above).
3. Change `rethinkdb_host` and `rethinkdb_port` if you have to.
4. Clone this repo using `git clone git@git.teonite.net:inspectr/inspectr.git.`
5. Build docker image for example: `docker build -t inspectr_cli .` in inspectr directory
6. Run your docker image (for example: `docker run -t -v $PWD/project:/code inspectr_cli`). 
Give your `inspectr.json` path instead of `$PWD/project`. If everything went ok report will be sent to rethinkDB and will be available in inspectr-dashboard.
7. If you're using RethinkDB docker container local, and you'll see error `ConnectionRefusedError: [Errno 111] Connection refused` you have to link RethinkDB container to alias which will be use in `inspectr.json` as rethinkdb_host parameter.
Example: `docker run -t --link inspectrdashboard_rethinkdb_1:db_link -v $PWD:/code inspectr_cli` where `inspectrdashboard_rethinkdb_1` is name of your RethinkDB container and `db_link` is parameter of `rethinkdb_host` in `inspectr.json`.
8. Your reports should be saved in RethinkDB. Use [inspectr-dashboard][dashboard] to see results in browser.

For more informations check FAQ at the end of README.


# Using inspectr locally

1. Run RethinkDB
2. Install Python virtual environment using, e.x. this code :
```
pip3 install virtualenv
virtualenv env
source env/bin/activate
```
3. Install InspectR CLI to your venv
`python setup.py install`
4. Place inspectr.json in your project directory (example above) and use command `inspectr`.
5. If you're inspecting JS app install (global) dependencies from docker/package.json.
6. If you're inspecting python app install pip requirements from `requirements.pip` and `docker/reporter-requirements.pip`

### Running

Simply go to your project and InspectR it! (remember that it must contain `inspectr.json`!)

```
cd /path/to/my/project
inspectr
```

Results should be visible in [InspectR Dashboard][dashboard]


# List of reporters

List of reporters:
```
{
      "project_name": "My Project",
    
      "reporters": [
        {
          "type": "flake8",
          "command": "flake8 apps/"
        },
        {
          "type": "unittest",
          "command": "manage.py test"
        },
        {
          "type": "coverage-py",
          "command": "coverage report"
        },
        {
          "type": "pytest",
          "command": "coverage run -m py.test"
        },
        {
          "type": "eslint",
          "command": "eslint src/"
        },
        {
          "type": "jasmine",
          "command": "karma start karma.conf.js --single-run"
        },
        {
          "type": "mocha",
          "command": "karma start karma.conf.js --single-run"
        },
        {
          "type": "karma-coverage",
          "command": null
        }
      ]
    }
```




# FAQ
 
1. If you have problems setting up Your RethinkDB instance, simply run
 it as Docker container:
`docker run -d -p 8080:8080 -p 28015:28015 -p 29015:29015 rethinkdb`

2. You have to run inspectr at your app first to create Collections in RethinkDB, otherwise you'll see error "Collection *reports* does not exist" 

3. If one of your reporters will fail check if it is installed.
For example if you have error
	```
	[FAIL] eslint src/ (eslint) FileNotFoundError: [Errno 2] No such file or directory: 'eslint'
	```
	Install eslint `npm install -g eslint`

4. If you're inspecting django app local, install django in your virtualenv.
5. If you're using Inspectr in docker and you want to connect with RethinkDB on localhost run `docker run -t --net="host" -v $PWD/project:/code inspectr_cli`

[rethink]: https://www.rethinkdb.com/
[dashboard]: https://git.teonite.net/inspectr/inspectr-dashboard
[eagle-L]: md_imgs/eagle-L.png "Eagle"
[eagle-R]: md_imgs/eagle-R.png "Eagle"
[star]: md_imgs/star.png "Star"
[example]: md_imgs/example.png "Example"


CHANGELOG
v 0.9.1
- add ability to specify project name and rethinkdb config by env variables
- add upload script to pypi

tests/mocks.py
# -*- coding: utf-8 -*-
from textwrap import dedent


flake8_output = dedent("""\
    apps/core/sensitivity_analysis.py:129:1: R701  'analyse_sensitivity' is too complex (44)
    apps/core/signals.py:18:1: R701  'updateNpvFromProject' is too complex (23)
    apps/core/utils.py:15:1: R701  'floatify' is too complex (19)
    apps/core/utils.py:116:1: R701  'duplicate' is too complex (22)
    apps/export/views.py:182:1: R701  'PortfolioPDFView' is too complex (14)
    apps/export/views.py:193:5: R701  'get_context_data' is too complex (14)

    """)

unittest_output = dedent("""\
    .F
    ======================================================================
    FAIL: test_npv (apps.core.tests.TestIndices)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/core/apps/core/tests.py", line 340, in test_npv
        self.assertAlmostEqual(npv, Decimal(self.index_results['npv']), 2)
    AssertionError: Decimal('158.60223620475019773') != Decimal('157') within 2 places

    ----------------------------------------------------------------------
    Ran 2 tests in 0.368s

    FAILED (failures=1)
    """)

coverage_output = dedent("""\
    Name                                                                                                  Stmts   Miss  Cover
    -------------------------------------------------------------------------------------------------------------------------
    conf/loggersettings.py                                                                                   11      4    64%
    tools/manage.py                                                                                           8      0   100%
    -------------------------------------------------------------------------------------------------------------------------
    TOTAL                                                                                                  4572   1673    63%
    """)

coverage_output_small = dedent("""\
    Name           Stmts   Miss  Cover
    ----------------------------------
    __init__.py        0      0   100%
    inspectr.py       16     16     0%
    parsers.py        43      0   100%
    reporters.py      21     21     0%
    utils.py          60     60     0%
    ----------------------------------
    TOTAL            140     97    31%
    """)

eslint_output = dedent("""\
    /home/jck/workspace/work/teonite/azoty/web-app/src/ng-app/app.js
    1:1   warning  Rule 'generator-star' was removed and replaced by: generator-star-spacing    generator-star
    50:32  error    'defaultLanguage' is not defined                                             no-undef

    ✖ 47 problems (5 errors, 42 warnings)

    """)

karma_output = dedent("""\
    26 06 2016 12:12:55.610:WARN [watcher]: Pattern "/home/jck/workspace/work/teonite/azoty/web-app/node_modules/sinon/pkg/sinon-timers.js" does not match any file.
    26 06 2016 12:12:55.618:WARN [watcher]: Pattern "/home/jck/workspace/work/teonite/azoty/web-app/dist/lib/angularjs/angular-locale_en-gb.js" does not match any file.
    26 06 2016 12:12:56.434:INFO [karma]: Karma v1.0.0 server started at http://localhost:9876/
    26 06 2016 12:12:56.434:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
    26 06 2016 12:12:56.440:INFO [launcher]: Starting browser PhantomJS
    26 06 2016 12:12:56.665:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket /#IoRwzZfZzsf3BNGbAAAA with id 9818241
    PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 SUCCESS (0.206 secs / 0 secs)
    -------------------|----------|----------|----------|----------|----------------|
    File               |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
    -------------------|----------|----------|----------|----------|----------------|
     js/               |     1.85 |        0 |        0 |     1.85 |                |
      app.js           |     1.85 |        0 |        0 |     1.85 |... 700,701,705 |
     js/ng-app/        |     4.63 |     0.54 |     1.08 |     4.72 |                |
      cms.js           |     4.63 |     0.54 |     1.08 |     4.72 |... 10554,10556 |
     js/portfolio-app/ |     6.06 |        0 |        0 |      6.1 |                |
      portfolio.js     |     6.06 |        0 |        0 |      6.1 |... 6,2433,2438 |
    -------------------|----------|----------|----------|----------|----------------|
    All files          |     4.82 |     0.43 |     0.78 |     4.91 |                |
    -------------------|----------|----------|----------|----------|----------------|


    """)

karma_coverage_summary_line = "All files          |     4.82 |     0.43 |     0.78 |     4.91 |                |"
karma_summary_success_line = "PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 SUCCESS (0.205 secs / 0.001 secs)"
karma_summary_fail_line = "PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 (1 FAILED) ERROR (0.21 secs / 0.001 secs)"


tests/fixtures.py
# -*- coding: utf-8 -*-
from textwrap import dedent


flake8_output = dedent("""\
    apps/core/sensitivity_analysis.py:129:1: R701  'analyse_sensitivity' is too complex (44)
    apps/core/signals.py:18:1: R701  'updateNpvFromProject' is too complex (23)
    apps/core/utils.py:15:1: R701  'floatify' is too complex (19)
    apps/core/utils.py:116:1: R701  'duplicate' is too complex (22)
    apps/export/views.py:182:1: R701  'PortfolioPDFView' is too complex (14)
    apps/export/views.py:193:5: R701  'get_context_data' is too complex (14)

    """)

unittest_output = dedent("""\
    .F
    ======================================================================
    FAIL: test_npv (apps.core.tests.TestIndices)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/core/apps/core/tests.py", line 340, in test_npv
        self.assertAlmostEqual(npv, Decimal(self.index_results['npv']), 2)
    AssertionError: Decimal('158.60223620475019773') != Decimal('157') within 2 places

    ----------------------------------------------------------------------
    Ran 2 tests in 0.368s

    FAILED (failures=1)
    """)

coverage_output = dedent("""\
    Name                                                                                                  Stmts   Miss  Cover
    -------------------------------------------------------------------------------------------------------------------------
    conf/loggersettings.py                                                                                   11      4    64%
    tools/manage.py                                                                                           8      0   100%
    -------------------------------------------------------------------------------------------------------------------------
    TOTAL                                                                                                  4572   1673    63%
    """)

coverage_output_small = dedent("""\
    Name           Stmts   Miss  Cover
    ----------------------------------
    __init__.py        0      0   100%
    inspectr.py       16     16     0%
    parsers.py        43      0   100%
    reporters.py      21     21     0%
    utils.py          60     60     0%
    ----------------------------------
    TOTAL            140     97    31%
    """)

eslint_output = dedent("""\
    /home/jck/workspace/work/teonite/azoty/web-app/src/ng-app/app.js
    1:1   warning  Rule 'generator-star' was removed and replaced by: generator-star-spacing    generator-star
    50:32  error    'defaultLanguage' is not defined                                             no-undef

    ✖ 47 problems (5 errors, 42 warnings)

    """)

eslint_output_noerrors = dedent("""\
    """)

eslint_output_custom = dedent("""\
    > react-boilerplate@3.0.0 lint:js /home/user/workspace/project
    > eslint .


""")


jasmine_output = dedent("""\
    26 06 2016 12:12:55.610:WARN [watcher]: Pattern "/home/jck/workspace/work/teonite/azoty/web-app/node_modules/sinon/pkg/sinon-timers.js" does not match any file.
    26 06 2016 12:12:55.618:WARN [watcher]: Pattern "/home/jck/workspace/work/teonite/azoty/web-app/dist/lib/angularjs/angular-locale_en-gb.js" does not match any file.
    26 06 2016 12:12:56.434:INFO [karma]: Karma v1.0.0 server started at http://localhost:9876/
    26 06 2016 12:12:56.434:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
    26 06 2016 12:12:56.440:INFO [launcher]: Starting browser PhantomJS
    26 06 2016 12:12:56.665:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket /#IoRwzZfZzsf3BNGbAAAA with id 9818241
    PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 SUCCESS (0.206 secs / 0 secs)
    -------------------|----------|----------|----------|----------|----------------|
    File               |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
    -------------------|----------|----------|----------|----------|----------------|
     js/               |     1.85 |        0 |        0 |     1.85 |                |
      app.js           |     1.85 |        0 |        0 |     1.85 |... 700,701,705 |
     js/ng-app/        |     4.63 |     0.54 |     1.08 |     4.72 |                |
      cms.js           |     4.63 |     0.54 |     1.08 |     4.72 |... 10554,10556 |
     js/portfolio-app/ |     6.06 |        0 |        0 |      6.1 |                |
      portfolio.js     |     6.06 |        0 |        0 |      6.1 |... 6,2433,2438 |
    -------------------|----------|----------|----------|----------|----------------|
    All files          |     4.82 |     0.43 |     0.78 |     4.91 |                |
    -------------------|----------|----------|----------|----------|----------------|


    """)

jasmine_output_integers = dedent("""\
    26 06 2016 12:12:55.610:WARN [watcher]: Pattern "/home/jck/workspace/work/teonite/azoty/web-app/node_modules/sinon/pkg/sinon-timers.js" does not match any file.
    26 06 2016 12:12:55.618:WARN [watcher]: Pattern "/home/jck/workspace/work/teonite/azoty/web-app/dist/lib/angularjs/angular-locale_en-gb.js" does not match any file.
    26 06 2016 12:12:56.434:INFO [karma]: Karma v1.0.0 server started at http://localhost:9876/
    26 06 2016 12:12:56.434:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
    26 06 2016 12:12:56.440:INFO [launcher]: Starting browser PhantomJS
    26 06 2016 12:12:56.665:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket /#IoRwzZfZzsf3BNGbAAAA with id 9818241
    PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 SUCCESS (0.206 secs / 0 secs)
    -------------------|----------|----------|----------|----------|----------------|
    File               |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
    -------------------|----------|----------|----------|----------|----------------|
     js/               |     1.85 |        0 |        0 |     1.85 |                |
      app.js           |     1.85 |        0 |        0 |     1.85 |... 700,701,705 |
     js/ng-app/        |     4.63 |     0.54 |     1.08 |     4.72 |                |
      cms.js           |     4.63 |     0.54 |     1.08 |     4.72 |... 10554,10556 |
     js/portfolio-app/ |     6.06 |        0 |        0 |      6.1 |                |
      portfolio.js     |     6.06 |        0 |        0 |      6.1 |... 6,2433,2438 |
    -------------------|----------|----------|----------|----------|----------------|
     All files         |      100 |      100 |      100 |      100 |                |
    -------------------|----------|----------|----------|----------|----------------|


    """)


mocha_success_output = dedent("""\
    02 07 2016 09:19:45.773:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket /#VePfZy9E04WvElXUAAAA with id 13721294
      selectLocationState
        ✔ should select the route as a plain JS object
      hooks
        getHooks
          ✔ given a store, should return all hooks
        helpers
          injectAsyncReducer
            ✔ given a store, it should provide a function to inject a reducer
          injectAsyncSagas
            ✔ given a store, it should provide a function to inject a saga

    =============================== Coverage summary ===============================
    Statements   : 85.31% ( 604/708 ), 54 ignored
    Branches     : 88.66% ( 258/291 ), 94 ignored
    Functions    : 73.86% ( 113/153 ), 22 ignored
    Lines        : 77.25% ( 258/334 )
    ================================================================================

    Finished in 0.08 secs / 0.034 secs

    SUMMARY:
    ✔ 4 tests completed

    """)


mocha_fail_output = dedent("""\
    Finished in 0.095 secs / 0.039 secs

    SUMMARY:
    ✔ 3 tests completed
    ✖ 1 test failed

    """)

karma_coverage_summary_line = "All files          |     4.82 |     0.43 |     0.78 |     4.91 |                |"
jasmine_summary_success_line = "PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 SUCCESS (0.205 secs / 0.001 secs)"
jasmine_summary_fail_line = "PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 (1 FAILED) ERROR (0.21 secs / 0.001 secs)"

pytest_output_fail = dedent("""\
    __________________________________________________________________________________________ test_coverage_pytest_reporter ___________________________________________________________________________________________

    monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x7f7fd9b752e8>

        def test_coverage_pytest_reporter(monkeypatch):
            monkeypatch.setattr(subprocess, 'Popen', mockpopen)

    >       report = coverage_pytest_reporter({}, [])
    E       NameError: name 'coverage_pytest_reporter' is not defined

    tests/test_reporters.py:80: NameError
    ======================================================================================= 4 failed, 38 passed in 0.10 seconds ========================================================================================
    """)

pytest_output_success = dedent("""\
    =============================================================================================== test session starts ================================================================================================
    platform linux -- Python 3.5.1, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
    rootdir: /home/jck/workspace/work/teonite/inspectr/inspectr/inspectr, inifile:
    plugins: mock-1.1
    collected 21 items

    tests/test_utils.py .....................

    ============================================================================================ 21 passed in 0.05 seconds =============================================================================================
    """)


radon_maintainability_output = dedent("""\
    apps/api/mixins.py - A
    apps/api/serializers.py - B
    apps/api/tests.py - A
    apps/api/index_views.py - A
    apps/api/__init__.py - A
    apps/api/urls.py - A
    apps/api/excel_exporter.py - B
    apps/api/model_views.py - C
    apps/api/portfolio_util.py - C
    apps/api/admin.py - A
    """)


radon_cyclomatic_complexity_output = dedent("""\
    inspectr/main.py
        F 36:0 run - B
        F 31:0 get_datetime - A
    inspectr/utils.py
        F 13:0 validate_and_parse_config - B
        F 38:0 init_db - A
        F 28:0 validate_connector_config - A
        F 77:0 print_command - A
        F 60:0 save_report - A
        F 73:0 colored - A
        F 84:0 print_command_ok - A
        F 91:0 print_command_fail - A
    inspectr/executor.py
        F 11:0 decode - A
        F 22:0 execute - A
        F 5:0 clean - A
    """)


coffeelint_output = dedent("""\
    ✗ src/app/services/GenericCaseActions.coffee
       ✗ #32: Line exceeds maximum allowed length. Length is 88, max is 80.
    ✓ src/app/services/LawyerProfile.coffee
    ✓ src/app/services/VersionFactory.coffee
    ✗ src/app/services/VindicationCase.coffee
       ✗ #6: Line exceeds maximum allowed length. Length is 82, max is 80.

    ✗ Lint! » 98 errors and 0 warnings in 41 files

    """)

tslint_output = dedent("""\
    src/app/views/personalization/step3.ts[18, 13]: The selector of the component "PersonalizationStep3View" should be named kebab-case (https://goo.gl/mBg67Z)
    src/app/views/personalization/step3.ts[17, 1]: Implement lifecycle hook interfaces (https://goo.gl/w1Nwk3)
    src/app/views/personalization/step3.ts[23, 14]: The name of the class PersonalizationStep3View should end with the suffix Component (https://goo.gl/5X1TE7)
    src/app/views/start.ts[2, 9]: Unused import: 'ActivatedRoute'
    src/app/views/start.ts[1, 25]: " should be '
    src/app/views/start.ts[2, 38]: " should be '
    """)


tests/test_utils.py
from pyrsistent import freeze, thaw
import rethinkdb as r
import pytest
from inspectr.utils import init_db, save_report, validate_and_parse_config, validate_connector_config


minimal_valid_config = freeze({
    "project_name": "Project",
    "rethinkdb_host": "localhost",
    "rethinkdb_port": 28015,
    "reporters": []
})

valid_reporter_config = freeze({
    'type': 'flake8',
    'command': ['flake8 app/']
})


def test_validate_minimal_config():
    parsed = validate_and_parse_config(minimal_valid_config)
    assert bool(parsed)
    assert parsed == minimal_valid_config


def test_validate_minimal_config_fail():
    for key in ['project_name', 'reporters']:
        config = minimal_valid_config.remove(key)
        with pytest.raises(SystemExit):
            validate_and_parse_config(thaw(config))


def test_config_with_reporters_success():
    config = minimal_valid_config.transform(['reporters'], lambda reporters: reporters.append(valid_reporter_config))

    parsed = validate_and_parse_config(thaw(config))
    assert bool(parsed)
    assert parsed == config


def test_config_with_reporters_fail():
    for key in ['type', 'command']:
        invalid_reporter = valid_reporter_config.remove(key)
        config = minimal_valid_config.transform(['reporters'], lambda reporters: reporters.append(invalid_reporter))
        with pytest.raises(SystemExit):
            validate_and_parse_config(thaw(config))


def test_init_db(monkeypatch, mocker):
    db_create_stub = mocker.stub('db_create')
    errors_stub = mocker.stub('errors')
    db_stub = mocker.stub('db')

    monkeypatch.setattr(r, 'db_create', db_create_stub)
    monkeypatch.setattr(r, 'errors', errors_stub)
    monkeypatch.setattr(r, 'db', db_stub)
    init_db('fake_connection', minimal_valid_config)

    db_create_stub.assert_called_once_with(minimal_valid_config['rethinkdb_db'])
    db_stub.assert_called_with(minimal_valid_config['rethinkdb_db'])


def test_save_report(monkeypatch, mocker):
    db_create_stub = mocker.stub('db_create')
    errors_stub = mocker.stub('errors')
    db_stub = mocker.stub('db')
    connect_stub = mocker.stub('connect')

    monkeypatch.setattr(r, 'db_create', db_create_stub)
    monkeypatch.setattr(r, 'errors', errors_stub)
    monkeypatch.setattr(r, 'db', db_stub)
    monkeypatch.setattr(r, 'connect', connect_stub)
    save_report({'project_name': 'fake_name'}, minimal_valid_config)

    db_create_stub.assert_called_once_with(minimal_valid_config['rethinkdb_db'])
    db_stub.assert_called_with(minimal_valid_config['rethinkdb_db'])
    connect_stub.assert_called_with(
        minimal_valid_config['rethinkdb_host'],
        minimal_valid_config['rethinkdb_port']
    )


inspectr/executor.py
import subprocess
import re


def clean(console_output):
    """Removes ANSI escape sequences from given console ouput"""
    ansi_escape = re.compile(r'\x1b[^m]*m')
    return ansi_escape.sub('', console_output)


def decode(binary_strings):
    result = []
    for bs in binary_strings:
        if isinstance(bs, str):  # sometimes we get regular strings here
            result.append(bs)
            continue
        result.append(clean(bs.decode('utf-8')))

    return result


def execute(command):
    """Executes given command with subprocess.Popen and returns it's stdout and stderr decoded"""
    if command is None:
        # some parsers depend only on previous commands output
        # and don't need to execute any command themselves
        return '', ''

    process = subprocess.Popen(command.split(' '), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    out, err = decode(process.communicate())
    return out, err


inspectr/parsers.py
# -*- coding: utf-8 -*-
import re


def flake8_parser(stdout, stderr, previous_reports=None):
    lines = stdout.split('\n')
    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': {
            'total_errors': max(0, len(lines) - 2)
        }
    }


def unittest_parser(stdout, stderr, previous_reports=None):
    """
    Example total_line: u'Ran 2 tests in 0.396s'
    Example fail status line: u'FAILED (failures=17)'
    """
    lines = stderr.split('\n')  # unittest pushes output to stderr
    total_line, status_line = lines[-4:-1:2]

    total_tests = int(total_line.split()[1])
    failed_tests = 0 if status_line.startswith('OK') else int(status_line.split('=')[1][:-1])
    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': {
            'total_tests': total_tests,
            'failed_tests': failed_tests,
            'passed_tests': total_tests - failed_tests,
        }
    }


def coverage_py_parser(stdout, stderr, previous_reports=None):
    lines = stdout.split('\n')
    summary_line = ' '.join(lines[-2].split())  # remove duplicate spaces
    total_statements, total_missing, coverage_percent = summary_line.split()[1:]

    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': {
            'total_statements': int(total_statements),
            'total_missing': int(total_missing),
            'coverage_percent': int(coverage_percent[:-1])
        }
    }


def pytest_parser(stdout, stderr, previous_reports=None):
    """
    Example fail summary_line: =========== 4 failed, 38 passed in 0.10 seconds =============
    Example success summary_line: ========= 21 passed in 0.05 seconds =========
    """
    lines = stdout.split('\n')
    summary_line = lines[-2].split()

    if summary_line[2].lower() == 'passed':
        passed_tests = int(summary_line[1])
        failed_tests = 0
    else:
        passed_tests = int(summary_line[3])
        failed_tests = int(summary_line[1])

    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': {
            'passed_tests': passed_tests,
            'failed_tests': failed_tests,
            'total_tests': passed_tests + failed_tests
        }
    }


def eslint_parser(stdout, stderr, previous_reports=None):
    lines = stdout.split('\n')
    eslint_summary_regex = r'^✖.*'
    summary_lines = [line for line in lines if re.search(eslint_summary_regex, line)]
    if not summary_lines:
        # no summary, no errors
        return {
            'stdout': stdout,
            'stderr': stderr,
            'summary': {
                'total_problems': 0,
                'total_errors': 0,
                'total_warnings': 0
            }
        }
    summary_line = summary_lines[-1].split()

    # example summary line: u'✖ 1767 problems (120 errors, 1647 warnings)'
    total_problems, total_errors, total_warnings = int(summary_line[1]), int(summary_line[3][1:]), int(summary_line[5])

    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': {
            'total_problems': total_problems,
            'total_errors': total_errors,
            'total_warnings': total_warnings
        }
    }


def extract_jasmine_summary(line):
    """
    Example SUCCESS karma summary line:
    PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 SUCCESS (0.205 secs / 0.001 secs)
    Exmaple FAIL karma summary line:
    PhantomJS 2.1.1 (Linux 0.0.0): Executed 1 of 1 (1 FAILED) ERROR (0.21 secs / 0.001 secs)
    """
    # get totals
    totals = line.split(' Executed ')[1].split(' ')
    executed_tests, total_tests = int(totals[0]), int(totals[2])

    # get failed
    if 'SUCCESS' in line:
        failed_tests = 0
    else:
        failed_tests = int(totals[3][1:])

    return {
        'total_tests': total_tests,
        'executed_tests': executed_tests,
        'failed_tests': failed_tests,
        'passed_tests': executed_tests - failed_tests
    }


def extract_karma_coverage_summary(line):
    """
    Example coverage summary line (table row):
    All files          |     4.82 |     0.43 |     0.78 |     4.91 |                |
    """
    totals = line.replace(' ', '').split('|')
    statements_percent, branches_percent, functions_percent, lines_percent = [float(t) for t in totals[1:5]]
    return {
        'statements_percent': statements_percent,
        'branches_percent': branches_percent,
        'functions_percent': functions_percent,
        'lines_percent': lines_percent
    }


def jasmine_parser(stdout, stderr, previous_reports=None):
    lines = stdout.split('\n')
    jasmine_summary_regex = r'\(.+\): Executed \d+ of \d+ (SUCCESS|\(\d+ FAILED\)( ERROR)?) \(.+ secs / .+ secs\)'

    jasmine_summary = None
    summary_lines = [line for line in lines if re.search(jasmine_summary_regex, line)]
    for line in summary_lines:
        jasmine_summary = extract_jasmine_summary(line)

    if jasmine_summary is None:
        raise ValueError('Karma summary line not found in input: %s' % stdout)

    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': jasmine_summary
    }


def karma_coverage_parser(stdout, stderr, previous_reports=None):
    reports = [report for report in previous_reports if report['type'] in ['jasmine', 'mocha']]
    for report in reports:
        lines = report['stdout'].split('\n')
        coverage_summary_regex = r'All files\s*\|\s*\d+(\.\d+)?\s*\|\s*\d+(\.\d+)?\s*\|\s*\d+(\.\d+)?\s*\|\s*\d+(\.\d+)?\s*\|\s*\|'

        coverage_summary = None
        summary_lines = [line for line in lines if re.search(coverage_summary_regex, line)]
        for line in summary_lines:
            coverage_summary = extract_karma_coverage_summary(line)

    if coverage_summary is None:
        raise ValueError('Did not find karma coverage summary line in previous reports')

    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': coverage_summary
    }


def extract_mocha_summary(lines):
    """
    Example mocha summary lines (both lines can be missing if no tests passed/failed):
    ✔ 3 tests completed
    ✖ 1 test failed
    """
    passes, fails = 0, 0
    for line in lines:
        if line and line[0] == '✔':
            passes = int(line.split()[1])
        elif line and line[0] == '✖':
            fails = int(line.split()[1])

    return {
        'total_tests': passes + fails,
        'passed_tests': passes,
        'failed_tests': fails
    }


def mocha_parser(stdout, stderr, previous_reports=None):
    lines = stdout.split('\n')
    mocha_summary_start_regex = r'^SUMMARY:$'

    mocha_summary = None
    summary_lines = [index for index, line in enumerate(lines) if re.search(mocha_summary_start_regex, line)]
    for line_index in summary_lines:
        mocha_summary = extract_mocha_summary(lines[line_index + 1:line_index + 3])

    if mocha_summary is None:
        raise ValueError('Karma summary line not found in input: %s' % stdout)

    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': mocha_summary
    }


def radon_maintainability_parser(stdout, stderr, previous_reports=None):
    """
    Example line: apps/api/urls.py - A
    """
    lines = stdout.split('\n')[:-1]
    summary = {'A': 0, 'B': 0, 'C': 0}
    for line in lines:
        maintainability = line.split(' ')[2]
        assert maintainability in ['A', 'B', 'C']
        summary[maintainability] += 1

    summary['total'] = sum(summary.values())
    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': summary,
    }


def radon_cyclomatic_complexity_parser(stdout, stderr, previous_reports=None):
    """
    Example line:     F 31:0 get_datetime - A
    """
    lines = stdout.split('\n')[:-1]
    summary = {'A': 0, 'B': 0, 'C': 0, 'D': 0, 'E': 0, 'F': 0}
    for line in lines:
        words = line.split(' ')
        if len(words) == 1:
            continue

        complexity = words[8]
        assert complexity in ['A', 'B', 'C', 'D', 'E', 'F']
        summary[complexity] += 1

    summary['total'] = sum(summary.values())
    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': summary,
    }


def coffeelint_parser(stdout, stderr, previous_reports=None):
    """
    Example summary line:
    ✗ Lint! » 98 errors and 0 warnings in 41 files
    """
    lines = stdout.split('\n')
    summary_line = lines[-3].split(' ')
    total_errors = int(summary_line[3])
    total_warnings = int(summary_line[6])
    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': {
            'total_problems': total_errors + total_warnings,
            'total_errors': total_errors,
            'total_warnings': total_warnings
        }
    }


def tslint_parser(stdout, stderr, previous_reports=None):
    lines = stdout.split('\n')
    return {
        'stdout': stdout,
        'stderr': stderr,
        'summary': {
            'total_errors': max(0, len(lines) - 1)
        }
    }
